x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java

```diff
@@ -0,0 +1,106 @@
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptorsIntersection;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+public record RemoteAccessAuthentication(Authentication authentication, Collection<BytesReference> roleDescriptorsBytesIntersection) {
+    public static final String REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY = "_remote_access_authentication";
+
+    public static void writeToContextAsRemoteAccessAuthentication(
+        final ThreadContext ctx,
+        final Authentication authentication,
+        final RoleDescriptorsIntersection roleDescriptorsIntersection
+    ) throws IOException {
+        ensureContextDoesNotContainRemoteAccessAuthentication(ctx);
+        ctx.putHeader(REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY, encode(authentication, roleDescriptorsIntersection));
+    }
+
+    public static RemoteAccessAuthentication readFromContext(final ThreadContext ctx) throws IOException {
+        return decode(ctx.getHeader(REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY));
+    }
+
+    public static Set<RoleDescriptor> parseRoleDescriptorsBytes(final BytesReference roleDescriptorsBytes) {
+        try (
+            XContentParser parser = XContentHelper.createParser(XContentParserConfiguration.EMPTY, roleDescriptorsBytes, XContentType.JSON)
+        ) {
+            final List<RoleDescriptor> roleDescriptors = new ArrayList<>();
+            parser.nextToken();
+            while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                parser.nextToken();
+                final String roleName = parser.currentName();
+                roleDescriptors.add(RoleDescriptor.parse(roleName, parser, false));
+            }
+            return Set.copyOf(roleDescriptors);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static String encode(final Authentication authentication, final RoleDescriptorsIntersection roleDescriptorsIntersection)
+        throws IOException {
+        // If we ever lift this restriction, we need to ensure that the serialization of each set of role descriptors to raw bytes is
+        // deterministic. We can do so by sorting the role descriptors before serializing.
+        assert roleDescriptorsIntersection.roleDescriptorsList().stream().noneMatch(rds -> rds.size() > 1)
+            : "sets with more than one role descriptor are not supported for remote access authentication";
+        final BytesStreamOutput out = new BytesStreamOutput();
+        out.setVersion(authentication.getEffectiveSubject().getVersion());
+        Version.writeVersion(authentication.getEffectiveSubject().getVersion(), out);
```

Comment:
This will write the version redundantly (`Authentication.writeTo()`) writes the version already. However, it feels better to be explicit here and not make assumptions based on the internals of `Authentication.writeTo()`

Reply:
Yes you're absolutely right. I was thinking of `Authentication.encode` -- anyhow, this definitely belongs here 😅